File: appinventor/components/src/com/google/appinventor/components/common/ColorSensorMode.java
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright 2020 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package com.google.appinventor.components.common;

import java.util.HashMap;
import java.util.Map;

/**
 * Defines a ColorSensorMode type used by the Lego Ev3 color sensor.
 */
public enum ColorSensorMode implements OptionList<String> {
  Reflected("reflected", 0),
  Ambient("ambient", 1),
  Color("color", 2);

  private final String value;
  private final int intValue;

  ColorSensorMode(String mode, int intMode) {
    this.value = mode;
    this.intValue = intMode;
  }

  public String toUnderlyingValue() {
    return value;
  }

  public Integer toInt() {
    return this.intValue;
  }

  private static final Map<String, ColorSensorMode> lookup = new HashMap<>();

  static {
    for (ColorSensorMode mode : ColorSensorMode.values()) {
      lookup.put(mode.toUnderlyingValue(), mode);
    }
  }

  public static ColorSensorMode fromUnderlyingValue(String mode) {
    return lookup.get(mode);
  }
}

File: appinventor/components-ios/src/ColorSensorMode.swift
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright © 2025 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// https://www.apache.org/licenses/LICENSE-2.0

@objc public class ColorSensorMode: NSObject, OptionList {
  @objc public static let Reflected = ColorSensorMode("reflected", 0)
  @objc public static let Ambient = ColorSensorMode("ambient", 1)
  @objc public static let Color = ColorSensorMode("color", 2)

  private static let LOOKUP: [String: ColorSensorMode] = generateOptionsLookup(Reflected, Ambient, Color)

  let value: String
  let intValue: Int32

  @objc private init(_ value: String, _ intValue: Int32) {
    self.value = value
    self.intValue = intValue
  }

  @objc public func toUnderlyingValue() -> AnyObject {
    return value as AnyObject
  }

  @objc public func toInt() -> Int32 {
    return intValue
  }

  @objc public static func fromUnderlyingValue(_ value: String) -> ColorSensorMode? {
    return LOOKUP[value]
  }
}
